### ISSUE TITLE
Getting error "The value for field "flag" is not defined" when running  PMML file using JPMML 1.4.9

### ISSUE THREAD

**Issue by poonamsinghtr** [jpmml/jpmml-evaluator#185]
Hi ,

I have a PMML file which is based on xgboost algorithm. When I am executing this file using JPMML : PMML executor version 1.4.9  in Java , I am getting error "The value for field "flag" is not defined" where flag is one of the output variable of my PMML file. 

I have been given this file from one of the source so I cant share the PMML file. If someone can help me .This is very urgent. 

Thanks in advance.

**Comment by vruusmann** [comment]
What software was used to generate this XGBoost PMML file? This info should be available in the `/PMML/Header/Application` element.

Is it JPMML-XGBoost?

> When I am executing this file using JPMML : PMML executor version 1.4.9

If you have a PMML file and a sample data record, then you should try executing this combination using the latest 1.4.15 command-line application (`org.jpmml.evaluator.EvaluationExample`, see the README file): https://github.com/jpmml/jpmml-evaluator/releases/tag/1.4.15

> I have been given this file from one of the source so I cant share the PMML file

But you can share the full stack trace, no?

Right now, I would interpret this error message that the model requires "flag" input field value, but your input data record does not provide it (ie. is incomplete).

**Comment by poonamsinghtr** [comment]
Thank you for the prompt reply. 

Its says R with version 1.4

In my PMML file , there is a tag MiningModel which has algorithm name  xgboost(thats how I have identified the algorithm. I know nothing about the PMML models/algorithm)

I am using PMML evaluator jar (java) which takes this PMML model and calls the ModelEvaluator.java class and then MiningModelEvaluator and so on. When I am debugging it , it throws exception missingValueException : "The value for field "flag" is not defined" in class OutputUtil.java at line no 145 (below lines of code)

if(!predictions.containsKey(targetName)){
throw new MissingValueException(targetName, outputField);}

This looks like to me the same issue -
https://github.com/jpmml/jpmml-evaluator/issues/126

You have answered it that we should be using R2PMML package rather than PMML 

Do you mean that we should be using R2PMML Java library rather than PMML library?

Or Is something wrong in the provided PMML file ?

Or R is used to generate the PMML file which is based on xgboost alogrithm  thats why something is wrong in the PMML file. So can I suggest my source to use something else to generate the PMML file which is more compatible for xgboost alogrithm ?

**Comment by vruusmann** [comment]
TLDR: If you upgrade from 1.4.9 to the latest 1.4.15, then this exception should go away.

The PMML specification is rather vague about field scoping rules. The JPMML-Evaluator version 1.4.9 is very strict about field scopes, whereas the 1.4.15 version is much more relaxed.

> Do you mean that we should be using R2PMML Java library rather than PMML library?

Don't use the legacy [pmml](https://cran.r-project.org/package=pmml) package, use my [r2pmml](https://cran.r-project.org/package=r2pmml) package instead.

Right now, you should be installing the r2pmml package from GitHub, because there's a very new 0.24.2 version available.

The r2pmml package generates XGBoost PMML files that are correct (provable integration test coverage), and 2-3x smaller in size. And they do field scoping very correctly.

For starters, see here: https://github.com/jpmml/jpmml-r/blob/master/src/test/R/xgboost.R

> Or Is something wrong in the provided PMML file ?

The legacy pmml package is generating garbage. Throw it away, and start from scratch.
